cpp/include/cuvs/neighbors/cagra.hpp

```diff
@@ -644,11 +657,136 @@ auto build(raft::resources const& res,
  */
 
 /**
- * @defgroup cagra_cpp_index_search CAGRA search functions
+ * @defgroup cagra_cpp_index_add_nodes CAGRA graph node addition functions
  * @{
+ * @brief Add new vectors to the index.
+ *
+ * This function expects an input array that contains both the old and new dataset vectors
+ * (new vector appended to the end). This allows explicit control of dataset allocation.
+ * If there are no specific requirements for dataset allocation, then it is recommended
+ * to use cagra::extend to add new vectors.
+ *
+ * See [cagra::extend](#cagra::extend) for usage example
+ *
+ * @tparam T data element type
+ * @tparam IdxT type of the indices
+ *
+ * @param[in] handle raft resources
+ * @param[in] input_updated_dataset_view updated dataset (initial + additional dataset)
```

Comment:
I'm really struggling with the need to expose this function through the public APIs, especially when the cagra index contains methods for the dataset to be attached or reattached after the fact. 

End users generally interact with these index APIs as a black box- that is, they invoke cuVS APIs to build and populate an index object and then they invoke functions to act upon that index object without ever having to care what's actually being done in the index object.

I understand the want to provide helper functions for power users but this isn't the correct way to provide these types of functions. If the only reason for exposing this function is because we want to provide a way for the user to give the dataset with the additional vectors appended then I think we need to put some additional thought into the existing "update_dataset" methods on the index and add an option to the extend params that affords more control to the user. 

The problem is that this function assumes the user knows or cares about a graph object, which is intentionally being hidden inside the index. If we remove the graph output, we are left with the extend() function and what I think should be an option for the user to specify whether they're attaching the whole detaset with the new vectors appended. Does that make sense?

Reply:
Yes, I agree that providing the APIs as a black box would be better for most users. And providing an option to specify whether the passed dataset is an entire one sounds good. An alternative way is to provide a function to expand the memory space (num vectors) for the dataset and graph in the CAGRA index. We allocate the host memory for the new dataset because we need to replace the old dataset with the new one on the device memory. To do this, we 1) copy the old dataset on the host memory, 2) deallocate the device memory, 3) allocate the new memory space on the device memory, and 4) copy the old dataset and additional dataset to the device memory (I noticed that we only need to allocate the host memory for the old dataset). It would be beneficial if there is a functionality to expand the memory space for new vectors because we don't need to reallocate the dataset memory space as above every time as long as the memory space is left.